xdsl/interactive/app.py

```diff
@@ -144,7 +150,14 @@ def __init__(
         file_path: str | None = None,
         input_text: str | None = None,
         pass_pipeline: tuple[tuple[type[ModulePass], PipelinePassSpec], ...] = (),
+        all_dialects: dict[str, Callable[[], Dialect]] = get_all_dialects(),
+        all_passes: dict[str, Callable[[], type[ModulePass]]] = get_all_passes(),
     ):
+        self.all_dialects = tuple((d_name, d) for d_name, d in all_dialects.items())
+        self.all_passes = tuple(
+            (p_name, p()) for (p_name, p) in sorted(all_passes.items())
+        )
```

Comment:
This is code smell to me, if someone already has the dialects and passes as a tuple, they would need to first build a dict to pass to the init, it's a shame to make it impossible to do the dumb/fast thing.

Reply:
I kind of get this, though to me the dictionary is the "correct" interface, in that it stops you assigning the same name to two different dialects/passes, whereas storing it as a tuple is really just an implementation detail.

Reply:
Note that `all_passes` has to have every pass actually called at some point as well. Unsure what the nicest interface is, realistically this is called once per project

Reply:
I agree in general, and this is where frozendict is nice, as it guarantees both unicity and frozenness, but in this case I think we'll be ok with the potential for errors in this way. In either case, this is a separate discussion from whether the init allows for passing the object to be stored directly or not

Reply:
Not dictionary, since it's mutable, so tuple for now, please. In theory, we could add a dependency to frozendict, but I'm not sure it's worth it for this PR.

Reply:
I meant just for the init function

Reply:
Ah, I don't really see the point, if it's just `tuple(all_dialects.items())` on the client side, and avoids `isinstance` checks in the init.

Reply:
I had a secret plan of making that change after you merge your PR :)

Reply:
Because it seems like too big a change to do both at the same time, but there's not a particularly good reason to make it eager at the app level.

Reply:
What's the solution then? 
- dictionaries with default and convert in init
- choice of dictionary/tuple and need isinstance checks
- only tuple but don't give defaults

Reply:
only tuple, don't give defaults

Reply:
Done, tests aren't pretty atm but I'm guessing you're hoping fine tune the passes they need in a future pr

Reply:
Realised I could have put the default for dialects, as this shouldn't be too expensive, let me know if you want it in

Reply:
I think the user needs to know they're doing something potentially very expensive, and to expose how to customize it for downstream users such as yourself